### PR TITLE
MEM-1075: Charge for table init

### PR DESCRIPTION
**Pull request by aakoshh**
Addresses https://github.com/filecoin-project/ref-fvm/issues/1075

The PR changes `charge_for_init` to also charge 8 bytes worth of memory expansion for the minimum number of table elements. 

See notes in the issue for why I don't think we can charge for growing the table right now. Here's the TL;DR version:

We are using `parity-wasm` to parse Wasm modules and inject our code e.g. for gas accounting, which then we write back into a modified Wasm module (binary) that Wasmtime compiles to native code. The `parity-wasm` project is essentially deprecated, archived, read-only - it's not being updated with newer Wasm 2.0 features. The `table-grow` instruction is newer than `parity-wasm` and would not be recognised by the parser if it was part of a smart contract code. Therefore we can't inject code that charges for it either. We'd either have to bring `parity-wasm` up to speed with newer op codes, or find a different library to use.

All we can do at the moment is allow modules to state their minimum number of table elements and charge for those.

### CODE REVIEW

File: fvm/src/gas/price_list.rs [hunk 1]
```diff
@@ -860,6 +860,15 @@ impl PriceList {
         self.wasm_rules.memory_expansion_per_byte_cost * grow_memory_bytes as i64
     }
 
+    /// Returns the gas required for initializing tables.
+    pub fn init_table_gas(&self, min_table_elements: u32) -> Gas {
+        // Each element reserves a `usize` in the table, so we charge 8 bytes per pointer.
+        // https://docs.rs/wasmtime/2.0.2/wasmtime/struct.InstanceLimits.html#structfield.table_elements
+        self.wasm_rules.memory_expansion_per_byte_cost
+            * (min_table_elements as i32)
```
Comment (Stebalien) [inline fvm/src/gas/price_list.rs:868]: Let's deal with overflow here.

File: fvm/src/gas/price_list.rs [hunk 2]
```diff
@@ -860,6 +860,15 @@ impl PriceList {
...
+            * (core::mem::size_of::<*const ()>() as i32)
```
Comment (Stebalien) [inline fvm/src/gas/price_list.rs:869]: We need this charge to be architecture independent, so let's just charge for 8 bytes (64 bits).

File: fvm/src/syscalls/mod.rs [hunk 3]
```diff
@@ -149,6 +154,19 @@ fn min_memory_bytes(module: &Module) -> crate::kernel::Result<usize> {
     }
 }
 
+/// Get the minimum number of table elements a module will use.
+///
+/// This relies on a few assumptions:
+///     * That we use the default value for `InstanceLimits::tables` and only allow 1 table.
+///     * That `Linker::command` will only allow them to be exported with the name "table".
```
Comment (Stebalien) [inline fvm/src/syscalls/mod.rs:161]: We'll have to re-evaluate in m2.2, but this is probably good enough.

**Review by Stebalien — COMMENTED** [review]
Given that this is a fixed cost that can't be controlled by the user and that we can't quite implement it correctly, I'd like to punt this to M2.2.

**Comment by Stebalien** [comment]
I can also just merge this (once we've fixed the platform-dependent issue) but that's something we can talk about.

File: fvm/src/gas/price_list.rs (hunk 2 above)
Comment (aakoshh) [reply fvm/src/gas/price_list.rs:869]: Thanks. I had the same thought but I wasn't sure if it's possible to run this with 32 bit systems.

File: fvm/src/gas/price_list.rs (hunk 1 above)
Comment (aakoshh) [reply fvm/src/gas/price_list.rs]: Doesn't the way multiplication is implemented deal with it? 

https://github.com/filecoin-project/ref-fvm/blob/master/fvm/src/gas/mod.rs#L141

File: fvm/src/gas/price_list.rs (hunk 2 above)
Comment (Stebalien) [reply fvm/src/gas/price_list.rs]: Well... Ok, it's not. But I'd still prefer to just hard-code this.

File: fvm/src/gas/price_list.rs (hunk 1 above)
Comment (Stebalien) [reply fvm/src/gas/price_list.rs]: Ah, I mean wrap around when casting to an i32.

Comment (aakoshh) [reply fvm/src/gas/price_list.rs]: Oh, good catch, that could have been a nasty vulnerability :facepalm: 

I added variants of `Gas::mul` to handle these so we don't have to keep casting.

**Review by Stebalien — APPROVED** [review]

**Pull request merged**